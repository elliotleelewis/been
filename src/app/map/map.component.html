<mgl-map
	logoPosition="bottom-right"
	[attributionControl]="false"
	[minZoom]="minZoom"
	[style]="'mapbox://styles/elliotleelewis/ckj1w9ro12g8h19mqhvsgcbkg'"
	[zoom]="[minZoom]"
>
	<mgl-vector-source
		id="countries"
		url="mapbox://mapbox.country-boundaries-v1"
	></mgl-vector-source>
	<mgl-layer
		id="visited-countries"
		type="fill"
		source="countries"
		sourceLayer="country_boundaries"
		[filter]="filter"
		[paint]="{ 'fill-color': '#fd7e14', 'fill-opacity': 0.7 }"
	></mgl-layer>
</mgl-map>
