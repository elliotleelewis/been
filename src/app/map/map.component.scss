:host {
	mgl-map {
		width: 100%;
		height: 100%;

		::ng-deep {
			canvas {
				outline: none;
			}
		}
	}
}
