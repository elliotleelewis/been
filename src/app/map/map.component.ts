import {
	ChangeDetectionStrategy,
	Component,
	HostBinding,
	Inject,
	type OnDestroy,
	type OnInit,
} from '@angular/core';
import { type Map } from 'mapbox-gl';
import { type Observable, distinctUntilChanged, fromEvent, map } from 'rxjs';
import { SubSink } from 'subsink';

import {
	type Coordinate,
	type NestedCoordinates,
	getCoordsCenter,
} from '../helpers';
import { CountriesService } from '../services/countries.service';

@Component({
	selector: 'app-map',
	templateUrl: './map.component.html',
	styleUrls: ['./map.component.scss'],
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MapComponent implements OnInit, OnDestroy {
	@HostBinding('class')
	class = 'flex flex-col';

	readonly darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
	readonly lightThemeUrl = 'mapbox://styles/mapbox/light-v11';
	readonly prefersDark = window.matchMedia('(prefers-color-scheme: dark)')
		.matches;

	map?: Map;
	center?: Coordinate;
	minZoom = 1.8;

	private _subs = new SubSink();

	constructor(
		@Inject(CountriesService) private countriesService: CountriesService,
	) {}

	static get prefersDark$(): Observable<boolean> {
		return fromEvent<MediaQueryListEvent>(
			window.matchMedia('(prefers-color-scheme: dark)'),
			'change',
		).pipe(
			map((event) => event.matches),
			distinctUntilChanged(),
		);
	}

	get countriesSelected$(): Observable<string[]> {
		return this.countriesService.countries$.pipe(
			map((countries) =>
				countries.filter((c) => c.selected).map((c) => c.iso3166),
			),
		);
	}

	ngOnInit(): void {
		this._subs.sink = this.countriesService.focus$.subscribe((country) => {
			const query = this.map?.querySourceFeatures('countries', {
				sourceLayer: 'country_boundaries',
				filter: ['==', ['get', 'iso_3166_1'], country],
			});
			const feature = query?.[0];
			if (feature) {
				const { coordinates } = feature.geometry as {
					coordinates?: NestedCoordinates;
				};
				if (coordinates) {
					this.center = getCoordsCenter(coordinates);
				}
			}
		});

		this._subs.sink = MapComponent.prefersDark$.subscribe((prefersDark) => {
			this.map?.setStyle(
				prefersDark ? this.darkThemeUrl : this.lightThemeUrl,
			);
		});
	}

	ngOnDestroy(): void {
		this._subs.unsubscribe();
	}

	onMapLoad(map: Map): void {
		this.map = map;
	}
}
