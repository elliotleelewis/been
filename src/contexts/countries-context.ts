import { createContext, useContext } from 'react';

import type { Country } from '../models/country';
import type { Region } from '../models/region';

export interface CountriesContextType {
	countries: readonly Country[];
	regions: readonly Region[];
	focus: Country | null;
	addCountry: (countryCode: string) => void;
	removeCountry: (countryCode: string) => void;
	clearCountries: () => void;
}

export const CountriesContext = createContext<CountriesContextType | undefined>(
	undefined,
);

export const useCountries = (): CountriesContextType => {
	const context = useContext(CountriesContext);
	if (!context) {
		throw new Error('useCountries must be used within a CountriesProvider');
	}
	return context;
};
