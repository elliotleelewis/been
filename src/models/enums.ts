export enum MapboxSourceKeys {
	countries = 'countries',
}

export enum MapboxLayerKeys {
	been = 'been',
	buildings = 'buildings',
}
