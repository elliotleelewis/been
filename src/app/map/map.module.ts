import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { MAPBOX_API_KEY, NgxMapboxGLModule } from 'ngx-mapbox-gl';

import { ENVIRONMENT } from '../../environments/environment';

import { MapComponent } from './map.component';

@NgModule({
	declarations: [MapComponent],
	imports: [BrowserModule, NgxMapboxGLModule],
	exports: [MapComponent],
	providers: [
		{
			provide: MAPBOX_API_KEY,
			useValue: ENVIRONMENT.apiKeyMapbox,
		},
	],
})
export class MapModule {}
