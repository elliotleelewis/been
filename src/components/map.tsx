import { centerOfMass, featureCollection } from '@turf/turf';
import { type FillExtrusionPaint, type FillPaint } from 'mapbox-gl';
import {
	type ReactNode,
	forwardRef,
	memo,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
} from 'react';
import {
	Layer,
	type MapRef,
	NavigationControl,
	Map as ReactMapGL,
	Source,
} from 'react-map-gl';

import { useCountries } from '../contexts/countries-context';
import { useMatchMedia } from '../hooks/use-match-media';
import { MapboxLayerKeys, MapboxSourceKeys } from '../models/enums';
import { type ForwardedRefFunction } from '../types/utils';

const apiKeyMapbox = import.meta.env['VITE_API_KEY_MAPBOX'] as
	| string
	| undefined;
const testMode = import.meta.env.MODE === 'test';
const darkThemeUrl = 'mapbox://styles/mapbox/dark-v11';
const lightThemeUrl = 'mapbox://styles/mapbox/light-v11';

const minZoom = 1.8;

export interface MapForwardedRef {
	isSourceLoaded: ForwardedRefFunction<MapRef['isSourceLoaded']>;
	querySourceFeatures: ForwardedRefFunction<MapRef['querySourceFeatures']>;
}

interface Props {
	header: ReactNode;
}

export const Map = memo(
	forwardRef<MapForwardedRef, Props>(({ header }, ref) => {
		const internalRef = useRef<MapRef>(null);

		const prefersDark = useMatchMedia('(prefers-color-scheme: dark)');

		const { countries, focus } = useCountries();

		const selectedCountries = useMemo(() => {
			return countries.filter((c) => c.selected).map((c) => c.iso3166);
		}, [countries]);

		useImperativeHandle(
			/* eslint-disable react-compiler/react-compiler -- This is a bug with the ESLint plugin, this is correct behaviour according to the React docs (see here: https://react.dev/reference/react/useImperativeHandle#exposing-a-custom-ref-handle-to-the-parent-component). */
			ref,
			() => ({
				isSourceLoaded: (
					...params: Parameters<MapRef['isSourceLoaded']>
				) => {
					return internalRef.current?.isSourceLoaded(...params);
				},
				querySourceFeatures: (
					...params: Parameters<MapRef['querySourceFeatures']>
				) => {
					return internalRef.current?.querySourceFeatures(...params);
				},
			}),
			/* eslint-enable react-compiler/react-compiler */
			[],
		);

		useEffect(() => {
			if (!focus || !internalRef.current) {
				return;
			}
			const { current: map } = internalRef;
			const features = map.querySourceFeatures(
				MapboxSourceKeys.countries,
				{
					sourceLayer: 'country_boundaries',
					filter: ['==', ['get', 'iso_3166_1'], focus],
				},
			);
			if (features.length > 0) {
				const { geometry } = centerOfMass(featureCollection(features));
				const [lat = 0, lon = 0] = geometry.coordinates;
				map.flyTo({ center: [lat, lon], zoom: minZoom });
			}
		}, [focus, internalRef]);

		const beenFilter = useMemo(
			() => ['in', ['get', 'iso_3166_1'], ['literal', selectedCountries]],
			[selectedCountries],
		);
		const beenPaint: FillPaint = useMemo(
			() => ({
				/* eslint-disable @typescript-eslint/naming-convention -- mapbox-gl has specific property names we must use */
				'fill-color': '#fd7e14',
				'fill-opacity': 0.6,
				/* eslint-enable @typescript-eslint/naming-convention */
			}),
			[],
		);

		const buildingsFilter = useMemo(() => ['==', 'extrude', 'true'], []);
		const buildingsPaint: FillExtrusionPaint = useMemo(
			() => ({
				/* eslint-disable @typescript-eslint/naming-convention -- mapbox-gl has specific property names we must use */
				'fill-extrusion-color': [
					'case',
					[
						'in',
						['get', 'iso_3166_1'],
						['literal', selectedCountries],
					],
					'#fd7e14',
					'#fd7e14',
				],
				'fill-extrusion-height': [
					'interpolate',
					['linear'],
					['zoom'],
					15,
					0,
					15.05,
					['get', 'height'],
				],
				'fill-extrusion-base': [
					'interpolate',
					['linear'],
					['zoom'],
					15,
					0,
					15.05,
					['get', 'min_height'],
				],
				'fill-extrusion-opacity': 0.6,
				/* eslint-enable @typescript-eslint/naming-convention */
			}),
			[selectedCountries],
		);

		return (
			<>
				{header}
				<ReactMapGL
					mapboxAccessToken={apiKeyMapbox ?? ''}
					mapStyle={prefersDark ? darkThemeUrl : lightThemeUrl}
					antialias
					attributionControl={false}
					logoPosition="bottom-right"
					minZoom={minZoom}
					ref={internalRef}
					testMode={testMode}
				>
					<NavigationControl />
					<Source
						id={MapboxSourceKeys.countries}
						type="vector"
						url="mapbox://mapbox.country-boundaries-v1"
					/>
					<Layer
						id={MapboxLayerKeys.been}
						type="fill"
						source={MapboxSourceKeys.countries}
						source-layer="country_boundaries"
						beforeId="national-park"
						filter={beenFilter}
						paint={beenPaint}
					/>
					<Layer
						id={MapboxLayerKeys.buildings}
						type="fill-extrusion"
						source="composite"
						source-layer="building"
						minzoom={15}
						filter={buildingsFilter}
						paint={buildingsPaint}
					/>
				</ReactMapGL>
			</>
		);
	}),
);
Map.displayName = 'Map';
