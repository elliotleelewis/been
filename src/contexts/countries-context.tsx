import {
	type FC,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { useLocalStorage } from '../hooks/use-local-storage';
import { type Country } from '../models/country';
import { type Region } from '../models/region';
import { regionalizer } from '../utils/regionalizer';

const COUNTRIES_STORAGE_KEY = 'APP_COUNTRIES';

interface CountriesContextType {
	countries: readonly Country[];
	regions: readonly Region[];
	focus: string | null;
	addCountry: (countryCode: string) => void;
	removeCountry: (countryCode: string) => void;
	clearCountries: () => void;
}

export const CountriesContext = createContext<CountriesContextType | undefined>(
	undefined,
);

export const useCountries = (): CountriesContextType => {
	const context = useContext(CountriesContext);
	if (!context) {
		throw new Error('useCountries must be used within a CountriesProvider');
	}
	return context;
};

interface Props {
	data: readonly Country[];
	children: ReactNode;
}

export const CountriesProvider: FC<Props> = ({ data, children }) => {
	const localStorage = useLocalStorage();
	const [selectedCountries, setSelectedCountries] = useState<string[]>(() => {
		const item = localStorage.getItem(COUNTRIES_STORAGE_KEY);
		return item ? (JSON.parse(item) as string[]) : [];
	});
	const [focus, setFocus] = useState<string | null>(null);

	useEffect(() => {
		localStorage.setItem(
			COUNTRIES_STORAGE_KEY,
			JSON.stringify(selectedCountries),
		);
	}, [selectedCountries, localStorage]);

	const countries = useMemo((): readonly Country[] => {
		return data.map((c) => ({
			...c,
			selected: selectedCountries.includes(c.iso3166),
		}));
	}, [data, selectedCountries]);

	const regions = useMemo((): readonly Region[] => {
		return regionalizer(countries);
	}, [countries]);

	const addCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			if (!prevCountries.includes(countryCode)) {
				setFocus(countryCode);
				return [...prevCountries, countryCode];
			}
			return prevCountries;
		});
	}, []);

	const removeCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			const newCountries = prevCountries.filter(
				(code) => code !== countryCode,
			);
			return newCountries;
		});
	}, []);

	const clearCountries = useCallback(() => {
		setSelectedCountries([]);
	}, []);

	const value = useMemo(
		() => ({
			countries,
			regions,
			focus,
			addCountry,
			removeCountry,
			clearCountries,
		}),
		[countries, regions, focus, addCountry, removeCountry, clearCountries],
	);

	return (
		<CountriesContext.Provider value={value}>
			{children}
		</CountriesContext.Provider>
	);
};
