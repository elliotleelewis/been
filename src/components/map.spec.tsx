import { render } from '@testing-library/react';
import { createRef } from 'react';
import { beforeAll, describe, expect, it, vi } from 'vitest';

import { CountriesContext } from '../contexts/countries-context';
import countriesJson from '../data/countries.json';
import { MapboxSourceKeys } from '../models/enums';

import { Map, type MapForwardedRef } from './map';

describe('Map', () => {
	beforeAll(() => {
		vi.spyOn(window, 'matchMedia').mockImplementation(
			() =>
				({
					matches: false,
					addEventListener: vi.fn(),
					removeEventListener: vi.fn(),
				}) satisfies Partial<MediaQueryList> as unknown as MediaQueryList,
		);
	});

	it('should render', () => {
		const result = render(
			<CountriesContext.Provider
				value={{
					countries: [],
					regions: [],
					focus: null,
					addCountry: vi.fn(),
					removeCountry: vi.fn(),
					clearCountries: vi.fn(),
				}}
			>
				<Map header={<div>Header!</div>} />
			</CountriesContext.Provider>,
		);

		expect(result.asFragment()).toMatchSnapshot();
	});

	it('should accept a ref', () => {
		const map = createRef<MapForwardedRef>();
		const result = render(
			<CountriesContext.Provider
				value={{
					countries: [],
					regions: [],
					focus: null,
					addCountry: vi.fn(),
					removeCountry: vi.fn(),
					clearCountries: vi.fn(),
				}}
			>
				<Map ref={map} header={<div>Header!</div>} />
			</CountriesContext.Provider>,
		);

		expect(map.current).toBeTruthy();
		expect(result.asFragment()).toMatchSnapshot();
	});

	it('should have geojson data for every country in dataset', async () => {
		const map = createRef<MapForwardedRef>();
		render(
			<CountriesContext.Provider
				value={{
					countries: [],
					regions: [],
					focus: null,
					addCountry: vi.fn(),
					removeCountry: vi.fn(),
					clearCountries: vi.fn(),
				}}
			>
				<Map ref={map} header={<div>Header!</div>} />
			</CountriesContext.Provider>,
		);

		await vi.waitUntil(
			() => map.current?.isSourceLoaded(MapboxSourceKeys.countries),
			{
				timeout: 5000,
			},
		);

		expect(map.current?.querySourceFeatures).toBeTruthy();
		for (const { iso3166, name } of countriesJson) {
			const features = map.current?.querySourceFeatures(
				MapboxSourceKeys.countries,
				{
					sourceLayer: 'country_boundaries',
					filter: ['==', ['get', 'iso_3166_1'], iso3166],
				},
			);
			expect.soft(features?.length, name).toBeGreaterThanOrEqual(1);
		}
	});
});
