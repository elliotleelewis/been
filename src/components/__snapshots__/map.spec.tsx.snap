// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map > it should render 1`] = `
<DocumentFragment>
  <div>
    Header!
  </div>
  <div
    style="position: relative; width: 100%; height: 100%;"
  />
</DocumentFragment>
`;
